### DISCUSSION TITLE
Why do json "database" libraries like Lowdb not work in API routes on Vercel/Netlify? Is this related to serverless functions?

### DISCUSSION THREAD

**Discussion by MadMaxMcKinney** [vercel/next.js#34936] [Help]
As I've done more research this seems to be related to running `lowdb` on Netlify and/or Vercel due to the API routes running as serverless functions. What I don't understand is _why_ this doesn't work. Why does this break on a live server but not locally? What is it about libraries like `lowdb` that don't make them compatible with how Netlify/Vercel handle Next's API routes? I want to understand the concepts that are happening here so I can find an alternative solution and/or make my own. I just don't want to use something else that works but not understand _why_.

For reference my code:

Here is my db init file (root/db/db.js)

```javascript
import {Low, JSONFileSync} from 'lowdb'

// Cluster DB Setup
const adapter = new JSONFileSync('cluster-db.json')
const clusterDB = new Low(adapter)

// Initialize if empty
clusterDB.read()
clusterDB.data ||= { clusters: [] }
clusterDB.write()

export {clusterDB}
```

And my only API route (root/pages/api/clusters.js)

```javascript
import {clusterDB} from '../../db/db'

export default async function handler(req, res) {

    await clusterDB.read()

    switch(req.method) {

        case 'POST':
            let newCluster = {severity: req.query.severity, comments: req.query.comments, date: req.query.date}
            clusterDB.data.clusters.push(newCluster)
            clusterDB.write()
            res.status(200).json({status: "Success", cluster: newCluster})
            break;

        case 'GET':
            if(clusterDB.data.clusters) {
                res.status(200).json(clusterDB.data.clusters)
            } else {
                res.status(404).json({status: "404"})
            }
            break;

    }

    res.status(200).json({test: "yay"})
}
```

lowdb library: [https://github.com/typicode/lowdb](https://github.com/typicode/lowdb)

**Comment by karandwivedi42** [comment]
https://vercel.com/guides/is-sqlite-supported-in-vercel

> SQLite needs a local file system on the server to store the data permanently when write requests are made. In a serverless environment, this central single permanent storage is not available because storage is ephemeral with serverless functions. As a function receives more concurrent traffic, the serverless environment will create new instances of the function and each instance will not be able to share the same storage.
